### PR TITLE
gh-58749: Removes the erroneous explanation regarding the restrictions of the global statement

### PR DESCRIPTION
**Pull request by bombs-kim**
* Issue: gh-58749

* [related mail chain](https://mail.python.org/pipermail/python-ideas/2012-April/014785.html)

As pointed out by @ncoghlan , CPython does not restrict a name in a global statement from reappearing in a [for](https://docs.python.org/dev/reference/compound_stmts.html#for) target list, [class](https://docs.python.org/dev/reference/compound_stmts.html#class) definition, function definition, or [import](https://docs.python.org/dev/reference/simple_stmts.html#import) statement

Additionally, CPython does not impose the same restrictions on [with](https://docs.python.org/dev/reference/compound_stmts.html#with) statements or [except](https://docs.python.org/dev/reference/compound_stmts.html#except) clauses, or or [variable annotations](https://docs.python.org/dev/glossary.html#term-variable-annotation).

The only real restriction is that a variable cannot be used or assigned to prior to its global declaration in the scope.

A few more notes on the changes
* I’ve even omitted the part about restrictions on formal parameters, although it's a real restriction. This is because 'a variable cannot be used or assigned to prior to its global declaration in the scope' already implies that restriction.
* I’ve revised the sentence to make it more similar to the nonlocal section.  Both explain a similar concept, and I believe the nonlocal explanation is clearer and more concise.
> The nonlocal statement applies to the entire scope of a function or
> class body. A :exc:`SyntaxError` is raised if a variable is used or
> assigned to prior to its nonlocal declaration in the scope.
* I removed the sentence, “The :keyword:global statement is a declaration which holds for the entire current code block,” as it is redundant with the following explanation: “The global statement applies to the entire scope of a function or class body.”

----
📚 Documentation preview 📚: https://cpython-previews--126523.org.readthedocs.build/

### CODE REVIEW

**Comment by bombs-kim** [comment]
As a side note, in #13232, @andresdelfino addressed the same section. However, it did not correct the erroneous explanation noted in #58749; it merely added cases to the existing list of name usages, which was already errorneous.

**Comment by andresdelfino** [comment]
> As a side note, in #13232, @andresdelfino addressed the same section. However, it did not correct the erroneous explanation noted in #58749; it merely added cases to the existing list of name usages, which was already errorneous.

Have you noticed the tests I added in the PR? Because I run them again, and verified that they still apply.

**Comment by bombs-kim** [comment]
@andresdelfino Yes! I've seen all the cases

Since this works, it's not accurate to describe that names listed in a :keyword:`global` statement must not be defined in a :keyword:`for` 

```
def for_global_before():
    global bar
    for bar in []:
        ...
```

Could you let me know if i'm missing something here?

**Comment by bombs-kim** [comment]
@ncoghlan thanks for noting an important point. I've checked that if mypy prevents certain usages of the same names after a global declaration. I've found no mention of such things in the documentation. Also, the following code passed the latest stable mypy (1.13.0)

```
def fn():
    global name

    with name:
        ...

    try:
        ...
    except name:
        ...

    class name:
        ...

    def name():
        ...

    var: name = 1

   # import name  # this will raise an error, but not because of the global declaration above
```

If you wish me to conduct more research on another type hints and static type checking tool, I would be happy to do that. However, I personally think, even if type checking tools prevent using the same names after a global declaration, it may not be a huge problem, since they are supposed to impose restrictions that the language spec does not have anyway.

The main problem this PR deals with is that there is a discrepancy btw the actual CPython implementation and the language spec document. I think it's better to get rid of the discrepancy either by changing the docs, or changing the compiler.

**Comment by ncoghlan** [comment]
Thanks for running that check, @bombs-kim. However, several of those checks aren't using their respective `as name` clauses, so they're not checking the name binding parts of those operations.

Re-reading the original issue, this change should also include a test suite addition (in https://github.com/python/cpython/blob/main/Lib/test/test_scope.py so the `global` and `nonlocal` test cases can go in the same file) that *ensures* these name bindings are all permitted.

Something like the following (using https://docs.python.org/3/reference/executionmodel.html#naming-and-binding to build the list of name binding operations to be checked):

```python
import contextlib
from types import SimpleNamespace

class TestGlobalNameBinding:
    # Ensure binding global names works as expected.
    # Genuine scope conflicts (such as names being both local and
    # nonlocal, or local and global) are covered in the compiler tests.
    # Note: does not check `async def`, `async for`, or `async with`.

    def test_assignment_statement(self):
        global _global_assignment_statement
        value = object()
        _global_assignment_statement = value
        self.assertIs(globals()["_global_assignment_statement"], value)
        del _global_assignment_expression

    def test_unpacking_assignment(self):
        global _global_unpacking_assignment
        value = object()
        _, _global_unpacking_assignment = [None, value]
        self.assertIs(globals()["_global_unpacking_assignment"], value)
        del _global_unpacking_assignment

    def test_assignment_expression(self):
        global _global_assignment_expression
        value = object()
        if (_global_assignment_expression := value):
            pass
        self.assertIs(globals()["_global_assignment_expression"], value)
        del _global_assignment_expression

    def test_iteration_variable(self):
        global _global_iteration_variable
        value = object()
        for _global_iteration_variable in [value]:
            pass
        self.assertIs(globals()["_global_iteration_variable"], value)
        del _global_iteration_variable

    def test_func_def(self):
        global _global_func_def
        def _global_func_def():
            pass
        value = _global_func_def
        self.assertIs(globals()["_global_func_def"], value)
        del _global_func_def

    def test_class_def(self):
        global _global_class_def
        class _global_class_def():
            pass
        value = _global_class_def
        self.assertIs(globals()["_global_class_def"], value)
        del _global_class_def

    def test_type_alias(self):
        global _global_type_alias
        type _global_type_alias = tuple[int, int]
        value = _global_type_alias
        self.assertIs(globals()["_global_type_statement"], value)
        del _global_type_alias

    def test_caught_exception(self):
        global _global_caught_exc
        try:
            1/0
        except ZeroDivisionError as _global_caught_exc:
            value = _global_caught_exc
        self.assertIs(globals()["_global_caught_exc"], value)
        del _global_caught_exc

    def test_caught_exception_group(self):
        global _global_caught_exc_group
        try:
            try:
                1/0
            except ZeroDivisionError as exc:
                raise ExceptionGroup("eg", [exc])
        except* ZeroDivisionError as _global_caught_exc_group:
            value = _global_caught_exc_group
        self.assertIs(globals()["_global_caught_exc_group"], value)
        del _global_caught_exc

    def test_enter_result(self):
        global _global_enter_result
        value = object()
        with contextlib.nullcontext(value) as _global_enter_result:
            pass
        self.assertIs(globals()["_global_enter_result"], value)
        del _global_enter_result

    def test_import_result(self):
        global _global_import_result
        value = contextlib
        import contextlib as _global_import_result
        self.assertIs(globals()["_global_import_result"], value)
        del _global_import_result

    def test_match(self):
        global _global_match
        value = object()
        match value:
            case _global_match:
                pass
        self.assertIs(globals()["_global_match"], value)
        del _global_match

    def test_match_as(self):
        global _global_match_as
        value = object()
        match value:
            case _ as _global_match_as:
                pass
        self.assertIs(globals()["_global_match_as"], value)
        del _global_match_as

    def test_match_seq(self):
        global _global_match_seq
        value = object()
        match (None, value):
            case (_, _global_match_seq):
                pass
        self.assertIs(globals()["_global_match_seq"], value)
        del _global_match_seq

    def test_match_map(self):
        global _global_match_map
        value = object()
        match {"key": value}:
            case {"key": _global_match_map}:
                pass
        self.assertIs(globals()["_global_match_map"], value)
        del _global_match_map

    def test_match_attr(self):
        global _global_match_attr
        value = object()
        match SimpleNamespace(key=value):
            case SimpleNamespace(key=_global_match_attr):
                pass
        self.assertIs(globals()["_global_match_attr"], value)
        del _global_match_attr
```

**Review by ncoghlan — CHANGES_REQUESTED** [review]
Blocking comment from the original issue discussion: the spec wording change should be accompanied by a test suite update that ensures the revised wording remains accurate.

For closing the original issue, the `nonlocal` wording also needs to be updated, but that can be done in a separate PR after the `global` change is resolved.

There's also a process question that I need to follow up before this can be merged (the level of review needed for language spec changes is generally higher than it was back when I first filed the linked issue, even when the spec change is fixing an error in the spec)

**Comment by ncoghlan** [comment]
In relation to the "global after use" error reporting behaviour, that's covered in https://github.com/python/cpython/blob/main/Lib/test/test_global.py

There is currently no `test_nonlocal.py` equivalent for `nonlocal` statements.

So while I suggested using `test_scope.py` above, another plausible way to arrange the tests for this behaviour would be:

* add the draft test case above to `test_global.py` (rewording the test module description accordingly) 
* add a new `test_nonlocal.py` that covered both the declaration-after-use equivalent checks for `nonlocal`, in addition to the `declaration-before-use` behaviour

The biggest advantage of this would be test discoverability, since the tests for these statements would be in files specifically named for them. The headers of each statement-specific test module could then refer off to `test_scope.py` as covering additional related behavioural tests.

**Comment by bombs-kim** [comment]
Hi @ncoghlan, I I’m updating you on the changes based on your feedback:
* I’ve chosen to update test_global.py rather than test_scope.py for the new tests.
* I've added the tests cases based on the code you provided. I've adjusted the existing tests to be consistent with the test
    * Tests for `test_caught_exception` and `test_caught_exception_group` have been slighted changed, as the original tests are not supposed to pass
* The module description has been updated to reflect the changes in tests and includes a reference to test_scope.py for related behaviors.

Also, the previous implementation details section has been removed. With the recent documentation update, the prior warning no longer applies.

Please let me know if you have additional suggestions.

**Review by ncoghlan — APPROVED** [review]

**Comment by ncoghlan** [comment]
Core dev consensus was that this is still a clear bug in the spec, so there's no need to get the SC directly involved.

File: Doc/reference/simple_stmts.rst
Comment (merwok) [inline Doc/reference/simple_stmts.rst:969]: This reads `The global`, so is missing a noun: `The global keyword`

Comment (bombs-kim) [reply Doc/reference/simple_stmts.rst:969]: Hey @merwok , thanks for the correction. I opened #126720 that reflects this feedback and handles some more mishaps.

**Pull request merged**
